clang/lib/Sema/SemaExprCXX.cpp

```diff
@@ -4728,6 +4728,10 @@ static bool CheckUnaryTypeTraitTypeCompleteness(Sema &S, TypeTrait UTT,
 
     return !S.RequireCompleteType(
         Loc, ArgTy, diag::err_incomplete_type_used_in_type_trait_expr);
+
+  // Only the type name matters, not the completeness, so always return true.
+  case UTT_SYCLMarkKernelName:
+    return true;
```

Comment:
Basically any type can be a kernel-name, as long as we can spell it.  The SYCL standard doesn't seem to require that 

    `kernel<int>(some_functor);`
    
is illegal.  In the cases of the unnamed lambda, it is obviously going to be a record type, but otherwise not necessarily.